aspnetcore/grpc/authn-and-authz.md

```diff
@@ -0,0 +1,143 @@
+---
+title: Authentication and authorization in gRPC for ASP.NET Core
+author: jamesnk
+description: Learn how to use authentication and authorization in gRPC for ASP.NET Core.
+monikerRange: '>= aspnetcore-3.0'
+ms.author: jamesnk
+ms.date: 06/07/2019
+uid: grpc/authn-and-authz
+---
+
+# Authentication and authorization in gRPC for ASP.NET Core
+
+By [James Newton-King](https://twitter.com/jamesnk)
+
+[View or download sample code](https://github.com/aspnet/AspNetCore.Docs/tree/master/aspnetcore/grpc/authn-and-authz/sample/) [(how to download)](xref:index#how-to-download-a-sample)
+
+## Authenticate users calling a gRPC service
+
+gRPC can be used with [ASP.NET Core authentication](xref:security/authentication/identity) to associate a user with each call. 
+
+The following is an example of `Startup.Configure` which uses gRPC and ASP.NET Core authentication:
+
+```csharp
+public void Configure(IApplicationBuilder app)
+{
+    app.UseRouting();
+    
+    app.UseAuthentication();
+    app.UseAuthorization();
+
+    app.UseEndpoints(endpoints =>
+    {
+        routes.MapGrpcService<GreeterService>();
+    });
+}
+```
+
+> [!NOTE]
+> The order in which you register the ASP.NET Core authentication middleware matters. Always call `UseAuthentication` and `UseAuthorization` after `UseRouting` and before `UseEndpoints`.
+
+Once authentication has been setup the user can be accessed in a gRPC service methods via the `ServerCallContext`.
+
+```csharp
+public override Task<BuyTicketsResponse> BuyTickets(
+    BuyTicketsRequest request, ServerCallContext context)
+{
+    var user = context.GetHttpContext().User;
+
+    // ... access data from ClaimsPrincipal ...
+}
+
+```
+
+### Bearer token authentication
+
+The client can provide an access token for authentication. The server validates the token and uses it to identify the user.
+
+On the server, bearer token authentication is configured using the [JWT Bearer middleware](/dotnet/api/microsoft.extensions.dependencyinjection.jwtbearerextensions.addjwtbearer).
+
+In the .NET gRPC client, the token can be sent with calls as a header:
+
+```csharp
+public bool DoAuthenticatedCall(
+    Ticketer.TicketerClient client, string token)
+{
+    var headers = new Metadata();
+    headers.Add("Authorization", $"Bearer {token}");
+
+    var request = new BuyTicketsRequest { Count = 1 };
+    var response = await client.BuyTicketsAsync(request, headers);
+
+    return response.Success;
+}
+```
+
+### Client certificate authentication
+
+A client could alternatively provide a client certificate for authentication. [Certificate authentication](https://tools.ietf.org/html/rfc5246#section-7.4.4) happens at the TLS level, long before it ever gets to ASP.NET Core. When the request enters ASP.NET Core the [client certificate authentication package](xref:security/authentication/certauth) allows you to resolve the certificate to a `ClaimsPrincipal`.
```

Comment:
Can you say something along the lines of you need to configure your server to accept/require a client certificate? From my reading, it almost sounds like all you need is the middleware which I found a little confusing